restart-dynflow/restart_dynflow.rake

```diff
@@ -0,0 +1,163 @@
+
+    def executor_worlds
+      return ForemanTasks.dynflow.world.coordinator.find_worlds(true).sort_by do |world|
+        GetProcessMem.new(world.data["meta"]["pid"]).gb
+      end.reverse
+    end
+
+    def waiting_processes(pids)
+      the_pids = pids.dup
+      begin
+        Timeout.timeout(90) do
+          while the_pids.any?
+            sleep 1
+            the_pids.each do |pid|
+              the_pids.delete(pid) unless Daemons::Pid.running?(pid)
+            end
+          end
+        end
+      rescue TimeoutError
+        puts "Pid #{the_pids.join(", ")} not exit within timeout killing them forcefully"
+        the_pids.each {|pid| Process.kill('KILL', pid)}
+        begin
+          Timeout.timeout(60) do
+            while the_pids.any?
+              sleep 1
+              the_pids.each do |pid|
+                the_pids.delete(pid) unless Daemons::Pid.running?(pid)
+              end
+            end
+          end
+        rescue TimeoutError
+          puts "Error: Timeout waiting pid #{the_pids.join(", ")} to exit"
+          exit 1
+        end
+      end
+    end
+
+    # Multiple monitor processes with the same appname will cause race condition
+    # when restarting the dynflow executors. Each monitor process will try to
+    # fork multiple executors and too many executors will be started. This code is
+    # to workaround this bug by killing duplicate monitor processes.
+    all_pids = `pgrep -f  'dynflow_executor.*_monitor'`.split("\n").map(&:strip).map(&:to_i)
+    if all_pids.empty?
+      puts "Error: Dynflow doesn't seem to be running"
+      exit 1
+    end
+
+    pid_dir = "/usr/share/foreman/tmp/pids/*"
+    wanted_pids = []
+    Dir.glob(pid_dir).select{|f| f =~ /dynflow_executor.*_monitor\.pid/}.each do |f|
+      wanted_pids.concat(`cat #{f}`.split("\n").map(&:strip).map(&:to_i))
+    end
+
+    if (all_pids & wanted_pids).empty?
+      puts "Error: Couldn't find any Dynflow executor monitor pid file in #{pid_dir}"
+      exit 1
+    end
+
+    unwanted_pids = all_pids - wanted_pids
+    unwanted_pids.each do |pid|
+      puts "Killing unwanted Dynflow executor monitor (pid #{pid})"
+      Process.kill('TERM', pid)
+    end
+    waiting_processes(unwanted_pids)
+
+    tried = 0
+    tries = 1
+    executors = []
+    all_worlds = executor_worlds
+    checklist = all_worlds.map(&:id)
+    while true
+      all_worlds.each do |world|
+        next unless checklist.include?(world.id)
+        pid = world.data["meta"]["pid"]
+        unless terminate_world?(world)
+          checklist.delete(world.id)
+          next
+        end
+        task_count = active_tasks(world).size
+        if task_count > 0
+          puts "Can't terminate world #{world.id}. Waiting for #{task_count} extensive tasks to finish" if tried < tries
+          next
+        end
+        mem = GetProcessMem.new(pid).gb.round(4)
+        puts "Terminating world #{world.id} (pid: #{pid}, memory used: #{mem}GB, limit: #{memory_limit}GB)"
```

Comment:
I like moving some of this output information (which is logged when called by cron) to the `terminate_world?(world)` check, so that the memory is logged even when the world isn't terminated (see other comment). It could be more useful for the user to monitor this way and understand why a restart wasn't performed.

Reply:
@wbclark Do you want to change the message here?

Reply:
Thinking on it some more, I think it's OK if we repeat some information in the termination message that was already output during the check.

Reply:
ya. the memory usage at the time that the script checks will not always equal to the memory usage when the process is terminating. e.g.  the script wait until the task is finish and do a second try.